pkg/recordset/recordset_test.go

```diff
@@ -1,15 +1,776 @@
+					StackName:   aws.String("cluster-foo-guest-recordsets"),
+					StackStatus: aws.String(cloudformation.StackStatusCreateComplete),
+				},
+				cloudformation.Stack{
+					StackName:   aws.String("cluster-baz-guest-recordsets"),
+					StackStatus: aws.String(cloudformation.StackStatusCreateComplete),
+				},
+			},
+			[]string{"cluster-foo-guest-recordsets", "cluster-baz-guest-recordsets"},
+		},
+
+		{
+			"do not update missing target stack",
+			[]cloudformation.Stack{
+				cloudformation.Stack{
+					StackName:   aws.String("cluster-foo-guest-main"),
+					StackStatus: aws.String(cloudformation.StackStatusCreateComplete),
+				},
+			},
+			nil,
+			nil,
+		},
+		{
+			"do not update missing source stack",
+			nil,
+			[]cloudformation.Stack{
+				cloudformation.Stack{
+					StackName:   aws.String("cluster-foo-guest-recordsets"),
+					StackStatus: aws.String(cloudformation.StackStatusCreateComplete),
+				},
+			},
+			nil,
+		},
+		{
+			"do not update missing source and target stacks",
+			nil,
+			nil,
+			nil,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.description, func(t *testing.T) {
+			logger, err := micrologger.New(micrologger.Config{IOWriter: ioutil.Discard})
+			if err != nil {
+				t.Fatalf("micrologger.New: %v", err)
+			}
+
+			sourceClient := newSourceWithStacks(tc.sourceStacks)
+			targetClient := newTargetWithStacks(tc.targetStacks)
+
+			c := &Config{
+				Logger:               logger,
+				Installation:         installation,
+				SourceClient:         sourceClient,
+				TargetClient:         targetClient,
+				TargetHostedZoneID:   zoneID,
+				TargetHostedZoneName: zoneName,
+			}
+			m, err := NewManager(c)
+			if err != nil {
+				t.Fatalf("NewManager: %v", err)
+			}
+
+			err = m.updateCurrentTargetStacks(tc.sourceStacks, tc.targetStacks)
+			if err != nil {
+				t.Fatalf("m.Sync: %v", err)
+			}
+
+			sort.Strings(targetClient.updatedStacks)
+			sort.Strings(tc.expectedUpdatedStacks)
+
+			if !reflect.DeepEqual(tc.expectedUpdatedStacks, targetClient.updatedStacks) {
+				t.Errorf("updated, expected %v  got %v", tc.expectedUpdatedStacks, targetClient.updatedStacks)
+			}
+		})
+	}
+}
+
+// TestUpdateCurrentTargetStacks_SourceStatuses tests Manager.updateCurrentTargetStacks
+//
+// Update is only allowed when source stack has status *_COMPLETE except DELETE_COMPLETE.
+func TestUpdateCurrentTargetStacks_SourceStatuses(t *testing.T) {
+	var (
+		installation = "installation"
+		zoneID       = "zoneID"
+		zoneName     = "zoneName"
+	)
+
+	tcs := []struct {
+		description  string
+		status       string
+		expectUpdate bool
+	}{
+		{
+			"update stack when source status is create complete",
+			cloudformation.StackStatusCreateComplete,
+			true,
+		},
+		{
+			"update stack when source status is update complete",
+			cloudformation.StackStatusUpdateComplete,
+			true,
+		},
+		{
+			"update stack when source status is rollback complete",
+			cloudformation.StackStatusRollbackComplete,
+			true,
```

Comment:
Does it make sense to create targed stack when source stack is rolled back?

Reply:
Very good point.

At first I though about rollback as being a successful state, but from experience I now know that when you create a stack it might happen that it get rolled back and therefore the status is similar to creation failed.

When rolling back from an update the stack would end up in it's previous successful state, so nothing would have changed.

With this in mind I now think that rollback status should be ignore as they either reflect no changes or a failure.

This simplifies the state transisions to:
* create/update when source stack is in state `CREATE_COMPLETE` or `UPDATE_COMPLETE`
* update when target stack is in `CREATE_COMPLETE`, `UPDATE_COMPLETE`, `ROLLBACK_COMPLETE`, `UPDATE_ROLLBACK_COMPLETE` or any `*_FAILED` state
* delete when source stack is in state `DELETE_COMPLETE`

Reply:
This makes sense with one caveat. I can't see how we can have anything to update on `ROLLBACK_COMPLETE`. This means the stack was rolled back and there are no resources created for this stack (unlike in `UPDATE_ROLLBACK_COMPLETE` case). This means we shouldn't have target stack created at all in `ROLLBACK_COMPLETE` IMO.

Reply:
Stack can't have `ROLLBACK_COMPLETE` status once it has `CREATE_COMPLETE` AIUI. But that shouldn't hurt to do it in that way.

Reply:
updated status tests to reflect those changed